### PR TITLE
Archive a problem

### LINKED ISSUE #849: Archive a problem
* Should not break old sections
* Hidden in all tasks and add task
* Show to admins and problem editors with a label
* Add a history item

### PR DESCRIPTION
**Pull request by IgorGoryany**
## PR includes

-   [x] Bug Fix
-   [x] Feature

## Related issues

Resolve #849

Related issues #878

## QA Instructions, Screenshots, Recordings

https://github.com/user-attachments/assets/893305f9-cd60-445e-b29d-e259fb65378c

https://github.com/user-attachments/assets/4dff8c80-1273-4e2f-bde5-16486c5adb27

### CODE REVIEW

File: src/modules/problemMethods.ts [hunk 1]
```diff
@@ -370,6 +382,20 @@ const update = async (data: UpdateProblem, authorId: number): Promise<Problem> =
...
     }
 
+    if (data.archived !== problem.archived) {
+        transactionOperations.push(
+            prisma.problemHistory.create({
+                data: {
+                    subject: 'archived',
+                    user: { connect: { id: authorId } },
+                    problem: { connect: { id: data.problemId } },
+                    previousValue: problem.archived ? 'in archive' : 'not in archive',
```
Comment (tsumo) [inline src/modules/problemMethods.ts:392]: Let's store this as boolean, and add translations during rendering

**Review by tsumo — CHANGES_REQUESTED** [review]

File: src/components/Problem/Problem.tsx [hunk 2]
```diff
@@ -51,6 +52,25 @@ export const Problem: FC<ProblemProps> = ({ problem }) => {
         destructive: true,
     });
 
+    const problemArchiveConfirmation = useConfirmation({
+        message: problem.archived ? tr('Remove problem from archive?') : tr('Archive problem?'),
+        onAgree: () =>
+            problemUpdateMutation
+                .mutateAsync({
+                    problemId: problem.id,
+                    archived: !problem.archived,
```
Comment (tsumo) [inline src/components/Problem/Problem.tsx:61]: You can pass only one field - `archived`, everything else stays the same

**Review by tsumo — CHANGES_REQUESTED** [review]

File: src/components/ProblemHistoryCard/ProblemHistoryCard.tsx [hunk 3]
```diff
@@ -46,8 +47,13 @@ export const ProblemHistoryCard: FC<ProblemHistoryCardProps> = ({ problemHistory
...
     const date = useDistanceDate(createdAt);
 
-    const beforeData = JSON.parse(JSON.stringify(previousValue, null, 4));
-    const afterData = JSON.parse(JSON.stringify(nextValue, null, 4));
+    let beforeData = JSON.parse(JSON.stringify(previousValue, null, 4));
+    let afterData = JSON.parse(JSON.stringify(nextValue, null, 4));
+
+    if (subject === 'archived') {
+        beforeData = previousValue === 'true' ? tr('in archive') : tr('not in archive');
+        afterData = nextValue === 'true' ? tr('in archive') : tr('not in archive');
```
Comment (asabotovich) [inline src/components/ProblemHistoryCard/ProblemHistoryCard.tsx:55]: I think it will be better to keep this logic on server side

Comment (asabotovich) [reply src/components/ProblemHistoryCard/ProblemHistoryCard.tsx:55]: Or keep in separate component
```tsx
<ArchivedChange from={beforeData} to={afterData} />
```

File: src/modules/problemMethods.ts [hunk 4]
```diff
@@ -59,6 +61,16 @@ const constructFindAllProblemsWhereFilter = async (
...
+    if (!admin && !problemEditor) {
+        if (where.AND && Array.isArray(where.AND)) {
+            where.AND.push({ OR: [{ authorId: userId }, { archived: false }] });
+        } else if (where.AND) {
+            where.AND.OR = [{ authorId: userId }, { archived: false }];
+        } else {
+            where.AND = { OR: [{ authorId: userId }, { archived: false }] };
+        }
+    }
+
```
Comment (asabotovich) [inline src/modules/problemMethods.ts:73]: I think we can reduce code difficulty if declare AND as Array at the beginnig of function

``` ts
const where = {
        AND: []
    } as Prisma.ProblemWhereInput;
```

File: src/components/Problem/Problem.tsx (hunk 2 above)
Comment (IgorGoryany) [reply src/components/Problem/Problem.tsx:61]: yes but then history items are created for changing all fields

File: src/components/HistoryArchivedTextChange/HistoryArchivedTextChange.tsx [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    from?: string | null;
+    to?: string | null;
+}
+
+export const HistoryArchivedTextChange: React.FC<HistoryChangeProps> = ({ from, to }) => {
+    const getTextContent = (text: string) => {
+        return text === 'true' ? tr('in archive') : tr('not in archive');
+    };
+    return (
+        <div className={s.HistoryArchivedTextChange}>
```
Comment (asabotovich) [inline src/components/HistoryArchivedTextChange/HistoryArchivedTextChange.tsx:17]: You can reuse HistoryTagsAndDifficultyTextChange, but rename it to abstract 'PlainTextChange'

File: src/components/ProblemHistoryCard/ProblemHistoryCard.tsx [hunk 6]
```diff
@@ -73,8 +73,8 @@ export const ProblemHistoryCard: FC<ProblemHistoryCardProps> = ({ problemHistory
                 </div>
             </div>
             {nullable(viewProblemHistoryDescription, () =>
-                subject === 'tags' || subject === 'difficulty' ? (
-                    <HistoryTagsAndDifficultyTextChange from={beforeData} to={afterData} />
+                subject === 'tags' || subject === 'difficulty' || subject === 'archived' ? (
+                    <ArchivedChange from={beforeData} to={afterData} />
```
Comment (asabotovich) [inline src/components/ProblemHistoryCard/ProblemHistoryCard.tsx:77]: For tags and subjects you should use PlainTextChange

File: src/components/ArchivedChange/ArchivedChange.i18n/ru.json [hunk 7]
```diff
@@ -0,0 +1,4 @@
+{
+    "in archive": "",
```
[the remark this answers is not in the document]
Comment (IgorGoryany) [reply src/components/ArchivedChange/ArchivedChange.i18n/ru.json]: Thanks!

**Review by asabotovich — APPROVED** [review]

**Review by tsumo — APPROVED** [review]

**Pull request merged**